### ISSUE TITLE
Parallelize code coverage builds

### ISSUE THREAD

**Issue by devjgm** [googleapis/google-cloud-cpp#3530]

**Comment by devjgm** [comment]
I think this issue needs a few more details explaining what needs to change. 

I found:

**Comment by coryan** [comment]
gcov corrupts the `.gcda` files if multiple executables are running in parallel, that is why we current **disable** parallel runs:

https://stackoverflow.com/questions/14643589/code-coverage-using-gcov-on-parallel-run

As you see there, there is a workaround:

https://gcc.gnu.org/onlinedocs/gcc-5.2.0/gcc/Cross-profiling.html

But then the problem becomes using a different value of `GCOV_PREFIX` for each binary, which *might* be possible using the right annotations for each test:

https://cmake.org/cmake/help/v3.6/prop_test/ENVIRONMENT.html

which would translate to adding a directive like this to each spot where we define a test:

```CMake
set_property(TEST ${test} ENVIRONMENT "GCOV_PREFIX=${something_unique}")
```

But this is all theory, somebody has to actually test it, and make sure it integrates with codecov.io and uploads the right things.

**Comment by coryan** [comment]
Just FYI, when serialized:

- The unit tests take 31s
- The bigtable integration tests take 28s
- The storage integration tests take 173s (!!) 
- The tests aganst production take 15s
- The remaining examples take about 2s

These time could be reduced by a factor of 4 if we parallelize them. With -spanner the numbers are even more significant, in this build the integration tests took **1800s** (!!!  -- but it varies a lot, sometimes it takes "as little as" 10 minutes to run the tests):

https://source.cloud.google.com/results/invocations/00c0c50f-08ca-4f97-a96c-e9dd8abf46b1/targets/cloud-cpp%2Fgithub%2Fgoogle-cloud-cpp-spanner%2Fmaster%2Fdocker%2Fcoverage-presubmit/log

In short, this build will become the bottleneck as we merge more projects into the monorepo.

**Comment by coryan** [comment]
I think I have bad news for this idea.  I prototyped using a different directory for each test, so they can be parallelized.  It works, but one then needs to merge the `.gcda` tests from each directory and that takes about 2 seconds per directory, we have about 200, so it takes 400 seconds to merge them back.

That is longer than we will probably save from parallelizing these.

For future reference the trick to splitting the output was to add this after each `add_test()`:

```
    string(SHA256 hash "${CMAKE_CURRENT_SOURCE_DIR}/${fname}" )
    set_property(TEST ${target} PROPERTY ENVIRONMENT  "GCOV_PREFIX=${CMAKE_BINARY_DIR}/gcda/${hash}")
```

And the merge step is:

```
for i in gcda/*; do echo "$(date -u) $i"; gcov-tool merge --output . . $i; done
```

**Comment by devjgm** [comment]
FWIW, closing this SGTM. If the serial gcov build ends up causing us pain, we'll open a new issue and revisit the problem at that point.
